Fix build errors from Geant4@10

drbenmorgan opened the pull request:
As reported in #828, the latest `develop` has a few compilation errors against Geant4 10.6 due to missing accessors. These are worked around here through use of the `G4VERSION_NUMBER` macro and direct/intermediate implementations:

- Missing `G4UIcommand::ConvertToLongInt()` is fixed by directly using `std::istringstream` as in the Geant4 implementation available from 10.7.
- Missing `GetTheta()`, `GetPhi()` accessors in `G4Trap` and `G4Para` calculate these parameters as in Geant4 11.0, using `GetSymAxis()` to obtain the needed inputs.
- Missing `GetAlpha...()` accessors in `G4Trap` simply use the `GetTanAlpha...()` accessors and `std::atan`.

The use of the direct Geant4 accessors is retained if the version supports them. Both implementations have been tested against Geant4 11, and produce the same results. Compilation against 10.6 works, but a few tests still fail:

```
The following tests FAILED:
	 81 - celeritas/ext/GeantImporter:FourSteelSlabs* (Failed)
	 83 - celeritas/ext/GeantImporter:OneSteelSphere.* (Failed)
	 98 - celeritas/global/AlongStep:Em3AlongStepTest.nofluct_nomsc (Failed)
	100 - celeritas/global/AlongStep:Em3AlongStepTest.msc_nofluct_finegrid (Failed)
	111 - celeritas/global/Stepper:TestEm3MscNofluct.* (Failed)
	151 - celeritas/user/Diagnostic:TestEm3* (Failed)
```

These need more investigation but on first look appear due to small (but above tolerance) differences in imported data, which is perhaps not surprising. I didn't look at trying to address these here until the compilation error fixes are reviewed and merged.

drbenmorgan committed:
Workaround no G4UIcommand::ConvertToLongInt in Geant4 < 10.7

Conversion from string to long int is only present in G4UIcommand from
Geant4 10.7, so to support 10.6 (e.g. for ATLAS) a workaround is
required.

Only use G4UIcommand::ConvertToLongInt when compiling against Geant 10.7
or newer. For older versions, use direct conversion via istringstream
as is done in the Geant4 implementation.

drbenmorgan committed:
Workaround G4Para/Trap accessors not available in Geant4 < 11

Direct accessors for theta/phi/alpha parameters on G4Para and G4Trap
were only added in Geant4 11.0, but we want to support 10.6 and newer
for ATLAS.

When compiling against Geant4 < 11, calculate theta, phi and alphas
using available symmetry axis and tan(alpha) accessors. Use direct
accessors otherwise.

sethrj commented on src/accel/SetupOptionsMessenger.cc:
```diff
@@ -54,7 +55,15 @@ struct UICommandTraits<T, std::enable_if_t<std::is_integral_v<T>>>
     static std::string to_string(T v) { return std::to_string(v); }
     static long from_string(G4String const& v)
     {
+        // Conversion to long int introduced in Geant4 10.7.0
+#if G4VERSION_NUMBER < 1070
+        G4long vl;
+        std::istringstream is(v);
+        is >> vl;
+        return vl;
+#else
```
For compiler preprocessing and regular program flow with if/else, I find it more readable to put the "usual case" first and then the "else" is the exceptional case (in this instance, for backward compatibility).

sethrj commented on src/celeritas/ext/g4vg/SolidConverter.cc:
```diff
@@ -561,18 +581,41 @@ auto SolidConverter::torus(arg_type solid_base) -> result_type
+    double const theta = std::atan2(tan_theta_sin_phi, tan_theta_cos_phi);
+    double const phi
+        = std::atan(std::sqrt(tan_theta_cos_phi * tan_theta_cos_phi
+                              + tan_theta_sin_phi * tan_theta_sin_phi));
```
Aside from the lack of scaling in this case, I think this is identical to the `para` above. Maybe it'd be worth making a helper function in the `namespace {` at the top of the file that returns a std::pair of theta/phi?

drbenmorgan replied:
I've refactored the whole calculation for both, as both are really unit vectors, or at least in the form `{A.tan(theta)cos(phi), A.tan(theta)sin(phi), A}` so the calculation the same irrespective of overall normalisation.

sethrj reviewed (approved):
Thanks so much for doing this! I have two small comments. I'll see about adding a Geant4@10.6 CI today.

drbenmorgan committed:
Put expected/normal code first in if preprocessor

Addresses review comments to have the "else" clause as the exceptional
case (here, for backward compatibility).

drbenmorgan committed:
Refactor theta/phi calculation for G4Para/G4Trap into free function

Use of Geant4 prior to 11.0 requires the theta/phi angles for G4Para
and G4Trap solids to be calculated from the components of the symmetry
axis vector. The encoding of the components is identical for both and
can thus be refactored into a common function.

Refactor calculation of theta/phi angles into a free function in
the anonymous namespace. Take symmetry vector/axis as argument,
returning theta, phi as a pair.
